docassemble/ALKilnTests/data/questions/test_loops.yml

```diff
@@ -0,0 +1,66 @@
+metadata:
+  title: Loop tests
+  description: Test loops like as with .there_is_another and .target_number
+---
+# Necessary to tell us what the sought var is on each page
+# Every interview that wants testing will need to have an element like this
+default screen parts:
+  post: |
+    <div data-variable="${ encode_name(str( user_info().variable )) }" id="trigger" aria-hidden="true" style="display: none;"></div>
+---
+objects:
+  - there_are_any_people: DAList.using(object_type=Individual)
+  - there_is_another_people: DAList.using(object_type=Individual, there_are_any=True)
+  - target_people: DAList.using(object_type=Individual, ask_number=True)
+---
+mandatory: True
+code: |
+  there_are_any_people.gather()
+  there_is_another_people.gather()
+  target_people.gather()
+  end1
+---
+id: there are any people
+generic object: DAList
+question: Are there any "${ x.object_name() }" people?
+yesno: x.there_are_any
+---
+id: is there another person
+generic object: DAList
+question: Are there more "${ x.object_name() }" people?
+yesno: x.there_is_another
+---
+id: target number
+question: How many "target_number" people are there?
+fields:
+  - Number of other people: target_people.target_number
+    input type: number
+---
+id: person name
+generic object: DAList
+question: Name of the "${ x[ i ].object_name() }" person
+fields:
+  - Name: x[i].name.first
+---
+id: end1
+event: end1
+question: The end of part 1!
```

Comment:
I actually intended to split the ones there into 'end1' and 'end2'. I can change that. I keep getting annoyed when we add pages into a test and have to update all related tests. I was hoping some format like this would allow us to add on to a test without having to update old tests. I wanted to think about it a little more. I can change it back for now if we want.

Reply:
> I was hoping some format like this would allow us to add on to a test without having to update old tests

Can you explain a bit more how this would avoid that? I can't picture it, as to reach two different endpoints, you'd have to have some choice affect that, and the only way I can picture that is the test making a specific radio button-like choice of "I'm running test 1". At that point, I wouldn't mind just keeping the different questions in different YAML files, even if it means some duplication on that side.

Reply:
Those are also alternatives. I'll outline what I'm thinking of anyway for clarity, though I'm not sure it's any better than the other options.

First what we have now:

First test we create:

```
Given I start...
And I do those first things
Then I reach the end screen
```

When we add the next feature, we change the first test:
```
Given I start...
And I do those first things
And I do the second things
Then I reach the end screen
```

With one test, that's fine, but not when we have tests with multiple paths (though granted I'm not sure how many of those we have, though I could envision others that test different inputs). With the new method:

First few tests we create with branching paths:
```
Given I start...
And I do those first things
Then I reach the 1st end screen
```

When we add the next feature, add a second test:
```
Given I start...
And I do those first things
Then I reach the 1st end screen
And I do the second things
Then I reach the 2nd end screen
```

We then don't have to change the first test to make sure it reaches the end.

Again, I'm not sure if this is the ideal way of handling this situation either. Maybe multiple test files is the way to go. Just clarifying what I meant.

Reply:
Hm, yeah, I don't think that's too useful. Test 2 is now running all of test 1, so what's the point of test 1?

* It might be useful to only be able to look at the title of the test that failed and guess what might have happened, but IMO:
    * that's rarely useful, but you have to dive into the code to see what the unexpected failure was anyway
    * if test 1 fails, test 2 will also fail; there's not a guaranteed order that people will see failures in (sometimes they might read from the top, but often I go to the bottom first, because it lists all of the failures, and I don't have to read through the "false failures" that are indistinguishable from real ones earlier in the logs), so you might see test 2 failed before you see test 1 failed, and wonder what could be going wrong with the 2nd thing (nothing, because it's the first thing). 

The main point of this comment is that `end1` doesn't make sense without an `end2`. IMO even if we still want that ending structure, I'd rather it be named `end_gathered` or something, I guess that's the practical takeaway from this discussion for now.

Reply:
> Test 2 is now running all of test 1, so what's the point of test 1?

As I said, this is in the case where test 1 is actually tests 1, 3, and 4. That is, where there are the same tests that test multiple paths on the same interview.

As I said, I was uncertain about this method anyway. I'll remove the `end1` name.

As a separate issue:

> I go to the bottom first, because it lists all of the failures, and I don't have to read through the "false failures" that are indistinguishable from real ones earlier in the logs

That's a problem I have too. Maybe we should add an internal way to tell the difference between internal false failures and true failures. Does that sound at all useful or will we end up just going to the bottom anyway?

Reply:
> As I said, this is in the case where test 1 is actually tests 1, 3, and 4

I didn't get that from your description, but my main takeaway was that end1 just wasn't a descriptive name.

> Does that sound at all useful or will we end up just going to the bottom anyway?

I would end up going to the bottom anyway.